### ISSUE TITLE
Request for "Scan pandoc md file for entries..." feature

### ISSUE THREAD

**Issue by jpk3333** [retorquere/zotero-better-bibtex#2079]
Thanks for BBT! I'd like to request a new feature: "Scan pandoc md file for entries...", which would be just like "Scan aux file for entries..." although it would scan an md file written with pandoc citations for the entries used in the file. I've learned that this can be done manually by exporting one's Zotero library to CSL JSON, and then running the getbib lua script described [here](https://twitter.com/pandoc_tips/status/1481910457145434113). But I'd love for this to be automated inside Zotero just as you've automated scanning an aux file. 

Thanks!

**Comment by retorquere** [comment]
This will have to assume that keys already exist in Zotero -- I can parse references from the pandoc AST, but I get no information on what bib/csljson files will be used by pandoc, so I can't import missing keys.

**Comment by jpk3333** [comment]
Absolutely. That's how it works now when Zotero scans an aux file, right? I'm thinking just that same functionality, but scanning an md file instead (with the Zotero bibtex keys being used to create the pandoc citations).

**Comment by retorquere** [comment]
The aux file has pointers to the bib files used to build it, and your can tell BBT to import those entries for which the key does not exist in Zotero.

**Comment by jpk3333** [comment]
Got it. I use Zotero exclusively, so it'd work for me. But I understand if you'd rather not build it for just my use case.

**Comment by retorquere** [comment]
It's almost done, it's a fairly small change, it's just that I can't replicate the entirety of the aux-scanning functionality.

**Comment by jpk3333** [comment]
Wonderful!

**Comment by jpk3333** [comment]
Thank you! I filed report id HGY5WG7H-euc.

The issue is that I can select AUX files, but md files remain greyed out:

<img width="194" alt="Screen Shot 2022-03-27 at 6 51 29 PM" src="https://user-images.githubusercontent.com/99686982/160306638-539876f3-7a48-4efa-b60f-e9a6577476a0.png">

<img width="278" alt="Screen Shot 2022-03-27 at 6 52 58 PM" src="https://user-images.githubusercontent.com/99686982/160306639-689dc02c-f958-4c30-be0d-c3717387318b.png">

**Comment by retorquere** [comment]
The pandoc executable must be in your PATH.

**Comment by jpk3333** [comment]
I'm pretty sure it is? I believe it's installed in /usr/local/bin/, and here is the output for `echo $PATH`:

`/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin:/Library/TeX/texbin`

**Comment by retorquere** [comment]
I can see in `HGY5WG7H-euc` that from the perspective of Zotero, your path is `/usr/bin:/bin:/usr/sbin:/sbin`.

**Comment by jpk3333** [comment]
What to do?

**Comment by retorquere** [comment]
Wow unless this has gotten better in the interim, it looks like [changing the PATH globally is a PITA on MacOS](https://stackoverflow.com/questions/36141784/sonarqube-local-script-in-intellij-cant-find-mvn-ioexception-no-such-directory/36161817#36161817). I can add `/usr/local/bin` on macos, because that is the default homebrew dir.

**Comment by retorquere** [comment]
Oh wait: https://mkyong.com/java/how-to-set-java_home-environment-variable-on-mac-os-x/

**Comment by jpk3333** [comment]
OK, so the issue concerns the $JAVA_HOME environment, and I might not need to add Pandoc to PATH? Or does setting the $JAVA_HOME environment somehow serve to do whatever is needed with Pandoc?

**Comment by retorquere** [comment]
Ah no, that's just meant to illustrate the way to modify environment variables in a way that GUI programs will see them. The variable in question that must be modified is `PATH` to add `/usr/local/bin`. If you're up to date and haven't reverted to bash, making the change in `~/.zshenv` should do it if I'm reading that page right.

**Comment by jpk3333** [comment]
OK, thanks. Yeah, so I'm ignorant at this level of detail, and my Unix buddy can't determine remotely why Pandoc/Zotero seems to be skipping the files that set /usr/local/bin and the TeX path (given that `echo $PATH` does show that `/usr/local/bin` is in my PATH.

**Comment by retorquere** [comment]
This is new to me too; I had to google these instructions. I'll experiment and report back.

**Comment by retorquere** [comment]
What does `echo $SHELL` return for you when entered in a command shell?

**Comment by jpk3333** [comment]
On account of privacy, I just DM'd some info to you on Twitter. Sorry if that's annoying.

**Comment by retorquere** [comment]
I don't think the shell matters in the end; this will do the trick:

```
sudo launchctl config user path $PATH
```

It will require a reboot, and it will only set the PATH-for-GUI-apps once, if you update your PATH later and you want that reflected in Zotero, you'd have to run this again.

Note that you should not take advice from strangers on the internet asking you to run `sudo <anything>`... and I only found this by applying google-fu, not any deep understanding of MacOS (which I do use). It Worked For Me, but you might want to consult a unix-guru pal. But on Linux (which I also use) this would have been way easier.

**Comment by retorquere** [comment]
```
sudo ln -s /usr/local/bin/pandoc /usr/bin/pandoc
```

will likely work for you. Still `sudo`, but at least what it's doing is straightforward and easily reversible.

**Comment by retorquere** [comment]
Wow, I see the same thing. Then I'm afraid the launchctl command is the only way I know how to make it work.

**Comment by jpk3333** [comment]
Ok, I used the launchctl command, rebooted, and it worked: I can now select md files. 

But...I don't see the new tag with the scanned references. This problem occurred with both md scanning and aux scanning. Error report #VWK8H6QB-euc.

**Comment by retorquere** [comment]
Did you scan to a tag, or to a collection?

**Comment by retorquere** [comment]
You can scan to a collection by right-clicking the collection. If the keys do not exist, nothing is created when tagging, although I suppose I could create a not reporting the error and tag that. When scanning to a collection, missing keys are reported in such a note.

**Comment by jpk3333** [comment]
Thanks, I didn't know about right-clicking to scan to a collection. Doing that, it still seems not to work. Unless I misunderstanding how this is supposed to work. No items at all are sent to that collection. All keys exist in my Zotero library, and I can compile from that library just fine using pandoc and setting my bibliography to use library I exported as Better CSL JSON.  Support log: UN3FWJ2D-euc

**Comment by retorquere** [comment]
If you go into Preferences -> Advanced -> Files and Folders -> Show Data Directory, the finder will open and show a `zotero` folder; inside that there should be a `better-bibtex` folder, and inside that a file names `list-citekeys.lua`. If you copy that file into the folder where your markdown source lives, and you run

```
pandoc --lua-filter list-citekeys.lua -t markdown -o citekeys.txt <filename of your markdown source>
```

(assuming that `citekeys.txt` **does not exist, or if it does, does not contain anything of importance to you**), what does `citekeys.txt` contain afterwards?

**Comment by jpk3333** [comment]
Brace yourself: no lua file at all in the better-bibtex folder. 

![Screen Shot 2022-03-29 at 4 01 47 PM](https://user-images.githubusercontent.com/99686982/160706483-96397380-43ea-4a3b-ba7f-380de1a792f7.png)

**Comment by retorquere** [comment]
can you install 2241, retry, then send a new log?

**Comment by jpk3333** [comment]
Error log SIIUK2GC-euc
![Screen Shot 2022-03-29 at 5 23 32 PM](https://user-images.githubusercontent.com/99686982/160716848-c2bfbb70-3b1a-4fd2-979c-1b81ff9cefde.png)

**Comment by retorquere** [comment]
I forgot to check in the filter file 🙄 . New build incoming.

**Comment by jpk3333** [comment]
Bummer, still no dice. Support log ID: IYI6TILA-euc

**Comment by retorquere** [comment]
Is the lua file present now? I don't see an attempt to install it in that log.

**Comment by retorquere** [comment]
Can I get a copy of that source file?

I take it you didn't actually type `<filename of your my source>`.

**Comment by retorquere** [comment]
You can either attach it here, or send it to emiliano.heyns@iris-advies.com

**Comment by retorquere** [comment]
You need to remove the existing lua file manually for now.

**Comment by jpk3333** [comment]
Still no dice. Reperformed citekeys.txt test w/ new lua filter and the file still ends up empty. Support log ID: 82CWEZKV-euc

**Comment by retorquere** [comment]
What pandoc version do you have?

As long as the manual run of pandoc doesn't put anything in citekeys.txt there is no need to send debug logs.

**Comment by jpk3333** [comment]
I have pandoc version 2.17.1.1. But I didn't see your earlier message about removing the existing lua file. Maybe that was the problem. What exactly do you want me to do with the lua file?

**Comment by jpk3333** [comment]
I have pandoc version 2.17.1.1. But I didn't see your earlier message about removing the existing lua file. Maybe that was the problem. What exactly do you want me to do with the lua file?

**Comment by retorquere** [comment]
A new build will drop in 10 minutes or so that will make sure you always have an up to date lua filter. If you don't want to wait, remove the lua file from the better bibtex folder and restart Zotero; BBT will put the new filter there. The upcoming build automates that.

**Comment by jpk3333** [comment]
That build worked!!!

**Comment by retorquere** [comment]
Took long enough. I've scheduled a new release.

**Comment by jpk3333** [comment]
I really appreciate your work!

**Comment by retorquere** [comment]
You have no idea how much I appreciate the interaction you offered here. It isn't always so pleasant.
